lib/src/inline_parser.dart

```diff
@@ -449,142 +743,156 @@ class StrikethroughSyntax extends TagSyntax {
+    i++;
+    if (i == parser.source.length) {
+      var text =
+          parser.source.substring(state.endPos, parser.pos).toLowerCase();
+      return _addNode(parser, state, label: text, endPos: i - 1);
+    }
+    var ch = parser.charAt(i);
 
-  /// Given that [match] has matched both a title and URL, creates an `<a>`
-  /// [Element] for it.
-  Element _createElement(InlineParser parser, Match match, TagState state) {
-    var link = getLink(parser, match, state);
-    if (link == null) return null;
+    // Maybe an inline link.
+    if (ch == $lparen) {
+      if (linkWalker.parseInlineLink()) {
+        // Huzzah, an inline link.
+        return true;
+      }
 
-    var element = new Element('a', state.children);
+      // At this point, we've matched `[...](`, but that `(` did not pan out to
+      // be an inline link. We must now check if `[...]` is simply a shortcut
+      // reference link.
+      var text =
+          parser.source.substring(state.endPos, parser.pos).toLowerCase();
+      return _addNode(parser, state, label: text, endPos: i - 1);
+    }
 
-    element.attributes["href"] = escapeHtml(link.url);
-    if (link.title != null) {
-      element.attributes['title'] = escapeHtml(link.title);
+    // A reference link.
+    if (ch == $lbracket) {
+      return linkWalker.parseReferenceLink();
     }
 
-    return element;
+    // Oops, perhaps just a simple shortcut reference link (`[...]`).
+    var text = parser.source.substring(state.endPos, parser.pos).toLowerCase();
+    return _addNode(parser, state, label: text, endPos: i - 1);
   }
 
-  /// Get the Link represented by [match].
-  ///
-  /// This method can return null, if the link is a reference link, and has no
-  /// accompanying link reference definition.
-  Link getLink(InlineParser parser, Match match, TagState state) {
-    if (match[3] != null) {
-      // Inline link like [foo](url).
-      var url = match[3];
-      var title = match[4];
-
-      // For whatever reason, Markdown allows angle-bracketed URLs here.
-      if (url.startsWith('<') && url.endsWith('>')) {
-        url = url.substring(1, url.length - 1);
-      }
-
-      return new Link(null, url, title);
-    } else {
-      String id;
-      String _contents() {
-        var offset = pattern.pattern.length - 1;
-        return parser.source.substring(state.startPos + offset, parser.pos);
-      }
-
-      // Reference link like [foo][bar].
-      if (match[1] == null) {
-        // There are no reference brackets ("shortcut reference link"), so infer
-        // the id from the contents.
-        id = _contents();
-      } else if (match[2] == '') {
-        // The id is empty ("[]") so infer it from the contents.
-        id = _contents();
-      } else {
-        id = match[2];
+  // Add a link node to [parser]'s AST.
+  //
+  // If [label] is present, the potential link is treated as a reference link.
+  // Otherwise, it is treated as an inline link.
```

Comment:
Good call. After refactoring, this class no longer needed `endPos`, at which point a reference Link is just the label. Since the fields would be mutually exclusive for inline and reference links, this has been severely refactored.